src/routes/reservation/index.ts

```diff
@@ -0,0 +1,252 @@
+}
+
+function createGetLockedRoomsCountHandler(beginDate: Date, endDate: Date) {
+  return (items: DynamoDB.Types.ItemList) => {
+    const dates = splitRange(beginDate, endDate);
+
+    let totalLockedRoomsCount = 0;
+
+    for (const date of dates) {
+      const lockedRoomsCount = items.reduce((acc, item) => {
+        const beginItemDate = new Date(String(item.beginDate.S));
+        const endItemDate = new Date(String(item.endDate.S));
+
+        if (beginItemDate <= date && date <= endItemDate) {
+          return acc + Number(item.bookedRoomsNumber.N);
+        }
+
+        return acc;
+      }, 0);
+
+      if (lockedRoomsCount > totalLockedRoomsCount) {
+        totalLockedRoomsCount = lockedRoomsCount;
+      }
+    }
+
+    return totalLockedRoomsCount;
+  };
+}
+
+export function getAvailableCountReservations() {
+  const dynamo = createDynamo();
+
+  const handler = async (event: awsx.apigateway.Request) => {
+    const propertyId = event.pathParameters!.id;
+    const beginDate = new Date(event.queryStringParameters!.beginDate);
+    const endDate = new Date(event.queryStringParameters!.endDate);
+
+    if (beginDate < new Date()) {
+      return buildApiResponse(500, {
+        message: 'Reservation date unavailable',
+      });
+    }
+
+    const property = await dynamo.getItem({
+      TableName: 'properties',
+      Key: {
+        id: { S: propertyId },
+      },
+    }).promise();
+
+    if (!property.Item) {
+      return buildApiResponse(404, {
+        message: 'Property not found',
+      });
+    }
+
+    const totalRoomsNumber = Number(property.Item!.totalRoomsNumber.N) || 1;
+
+    const params = {
+      TableName: 'reservation',
+      IndexName: 'reservation-by-property-id',
+      KeyConditionExpression: 'propertyId = :propertyId',
+      ExpressionAttributeValues: {
+        ':propertyId': { S: propertyId },
+      },
+      Limit: 1000,
+    };
+
+    const availableRoomsCount = await query(dynamo, params)
+      .then(createGetLockedRoomsCountHandler(beginDate, endDate))
+      .then((lockedRoomsCount) => totalRoomsNumber - lockedRoomsCount);
+
+    return buildApiResponse(200, { availableRoomsCount });
+  };
+
+  return add500Handler(handler);
+}
+
+export function createReservation() {
+  const dynamo = createDynamo();
+
+  const handler = async (event: awsx.apigateway.Request) => {
+    const customerId = getUserId(event);
+
+    const body = parseBody(event);
+
+    const propertyId = body.propertyId;
+    const bookedRoomsNumber = +body.bookedRoomsNumber;
+    const beginDate = new Date(body.beginDate);
+    const endDate = new Date(body.endDate);
+
+    const property = await dynamo.getItem({
+      TableName: 'properties',
+      Key: {
+        id: { S: propertyId },
+      },
+    }).promise();
+
+    if (!property.Item) {
+      return buildApiResponse(404, {
+        message: 'Property not found',
+      });
+    }
+
+    const totalRoomsNumber = Number(property.Item!.totalRoomsNumber.N) || 1;
```

Comment:
If we use the propertyModel.ts, we will receive a proper totalRoomsNumber, we will not need this fallback to 1 here